### PR TITLE
OverlappingFieldsCanBeMerged validation rule

### PR DESCRIPTION
**Pull request by crisu83**

### CODE REVIEW

**Comment by crisu83** [comment]
There is still one test case that fails, but I marked it as skipped and created a separate issue (#69) for it to be able to move on to the next rule.

**Comment by hugovk** [comment]
> and created a separate issue (#69) for it 

Do you mean https://github.com/digiaonline/graphql-php/issues/76?

File: src/Validation/Conflict/ComparisonContext.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    protected $fieldMap = [];
+
+    /**
+     * @var array
+     */
+    protected $fragmentNames = [];
+
+    /**
+     * @var array|Conflict[]
```
Comment (Jalle19) [inline src/Validation/Conflict/ComparisonContext.php]: `array` is superflous here

File: src/Validation/Conflict/ComparisonContext.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    public function hasConflicts(): bool
+    {
+        return !empty($this->conflicts);
+    }
+
+    /**
+     * @return array|Conflict[]
+     */
+    public function getConflicts()
```
Comment (Jalle19) [inline src/Validation/Conflict/ComparisonContext.php]: Could have static type-hint

File: src/Validation/Conflict/FieldContext.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+namespace Digia\GraphQL\Validation\Conflict;
+
+use Digia\GraphQL\Language\Node\FieldNode;
+use Digia\GraphQL\Type\Definition\CompositeTypeInterface;
+use Digia\GraphQL\Type\Definition\Field;
+
+class FieldContext
+{
+    /**
+     * @var CompositeTypeInterface|null
```
Comment (Jalle19) [inline src/Validation/Conflict/FieldContext.php]: `?ComposerTypeInterface`

File: src/Validation/Conflict/FieldContext.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+     */
+    protected $parentType;
+
+    /**
+     * @var FieldNode
+     */
+    protected $node;
+
+    /**
+     * @var Field|null
```
Comment (Jalle19) [inline src/Validation/Conflict/FieldContext.php]: `?Field`

File: src/Validation/Conflict/FieldContext.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        FieldNode $node,
+        ?Field $definition = null
+    ) {
+        $this->parentType = $parentType;
+        $this->node       = $node;
+        $this->definition = $definition;
+    }
+
+    /**
+     * @return CompositeTypeInterface|null
```
Comment (Jalle19) [inline src/Validation/Conflict/FieldContext.php]: Same for these as for the class properties

**Comment by Jalle19** [comment]
@crisu83 would it perhaps be a good idea to use https://github.com/php-ds/polyfill instead of creating custom Map and Pair types?

File: src/Validation/Conflict/FieldContext.php (hunk 3 above)
Comment (crisu83) [reply src/Validation/Conflict/FieldContext.php]: I had it like this, but I got errors for it. Can't remember what complained about them.

**Comment by crisu83** [comment]
@hugovk yes.

**Comment by crisu83** [comment]
@Jalle19 I wouldn't introduce that dependency because we still have to implement the `PairSet` or something similar to be able to keep track of the fragments checked.

**Pull request merged**